headers.c

```diff
@@ -390,9 +390,13 @@ GC_INNER struct hblk * GC_prev_block(struct hblk *h)
             } else if (IS_FORWARDING_ADDR_OR_NIL(hhdr)) {
                 j -= (signed_word)hhdr;
             } else {
+	      #if !defined(__CHERI_PURE_CAPABILITY__)
                 return((struct hblk *)
                           (((bi -> key << LOG_BOTTOM_SZ) + j)
                                << LOG_HBLKSIZE));
+	      #else
+                return((struct hblk *)hhdr->hb_block);
```

Comment:
Is there a reason why the bit-shifting dance doesn't make sense here?

Reply:
The capability that is returned from here is being utilized to coalesce and replace memory. The arithmetic pointer generation does not give us a capability with provenance. 
When this free block is then utilized, it will throw a SIGPROT.